cheatsheets/Password_Storage_Cheat_Sheet.md

```diff
@@ -1,567 +1,153 @@
 
-2. Load and use new protection scheme
-    1. Load a new, stronger credential protection scheme (See next section on: Upgrading your existing password hashing solution)
-    2. Include version information stored with form
-    3. Set 'tainted'/'compromised' bit until user resets credentials
-    4. Rotate any keys and/or adjust protection function parameters such as work factor or salt
-    5. Increment scheme version number
+In some cases, it may be possible to increase the work factor of the hashes without the original password, although this is not supported by common hashing algorithms such as Bcrypt and PBKDF2.
 
-3. When user logs in:
-    1. Validate credentials based on stored version (old or new); if older compromised version is still active for user, demand 2nd factor or secret answers until the new method is implemented or activated for that user
-    2. Prompt user for credential change, apologize, & conduct out-of-band confirmation
-    3. Convert stored credentials to new scheme as user successfully log in
+## Maximum Password Lengths
 
-## Upgrading your existing password hashing solution
+Some hashing algorithms such as Bcrypt have a maximum length for the input, which can [vary between different implementations](https://security.stackexchange.com/questions/39849/does-bcrypt-have-a-maximum-password-length) of the algorithm. Where Bcrypt is used, a maximum length of 50 characters should be enforced on the input to cover all known implementations.
 
-The above guidance describes how to do password hashing correctly/safely. However, it is very likely you'll be in a situation where you have an existing solution you want to upgrade. This [article](https://veggiespam.com/painless-password-hash-upgrades/) provides some good guidance on how to accomplish an upgrade in place without adversely affecting existing user accounts and future proofing your upgrade so you can seamlessly upgrade again (which you eventually will need to do).
+Additionally, due to how computationally expensive modern hashing functions are, if a user can supply very long passwords then there is a potential denial of service vulnerability, such as the one published in [Django](https://www.djangoproject.com/weblog/2013/sep/15/security/) in 2013.
 
-# Argon2 usage proposal in Java
+In order to protect against both of these issues, a maximum password length should be enforced. This should be 50 characters for Bcrypt, and at least 64 characters for other algorithms. There are two main ways that this can be done:
 
-The objective is to propose a example of secure usage/integration of the Argon2 algorithm in Java application to protect password when stored.
+- Restrict the maximum length in a password policy.
```

Comment:
We need to give some advice on the limit too, now? Like yeah, set the maximum length, but if a developer is reading this, they won't actually have a value. In the last version it was set to 160. For me, anything above 100 and less than 200 is an acceptable value.

Reply:
A few lines above, there is:

>   In order to protect against both of these issues, a maximum password length should be enforced. This should be 50 characters for Bcrypt, and at least 64 characters for other algorithms. There are two main ways that this can be done:

Do you want anything beyond that?

Reply:
At least 64 chars, but what about the actual maximum bound?
Reader: Oh at least 64, but yeah it can take 2k, since that is the same limit we have for parameters.
And that's how they were DoSsed 😛 Keep that text as is, and just add an upper bound limit in the bullet points.

Reply:
Updated to `between 64 and 128 characters`. 128 is slightly arbitrary - but there's not real difference in security between 50/64/128/256 chars.

Happy for some other thoughts on this.

Reply:
I agree that they are good for _now_. I don't mind that at all 😄